Remote log sync (WIP)

cincodenada opened the pull request:
This is a WIP, but I'm dogfooding it right now. Currently disabled by default, activates once you fill in `$remote_id` in settings. Throwing up this PR cause it's in MVP stage and thus good enough for some preliminary feedback.

So what's it do? Well, with the (rather generous) assumption that you have somewhere you can `scp` to, this implements a very basic remote-sync and backup capability for TagTime logs. I added two extra bits into `launch.pl` (with concordant additions to `settings.pl`) that accomplish the following:

- After any successful `launch.pl`, TagTime copies its log up to a remote server, tagged with `$remote_id`, a user-specified machine-specific string to distinguish it from any other instances, so that any instance can access any other instance's log.

- Then, when launching, if there's a gap between the last logged ping and now, before filling it in with RETRO, it tries to pull down the latest (by modification time) log from the specified remote server, which was presumably previously uploaded by a different machine. It fills in the gap with pings from that file as well as it can, updates `$nxtping` taking the new pings into account, and then proceeds as usual.

I'll find out soon enough (and let y'all know) if I've missed anything obvious, but I believe this should accomplish my goal of being able to hop between computers without any manual logfiddling, and as a bonus, would even automatically recover from your log (or TagTime folder) somehow being wiped.

There's a small gotcha currently: if you are on computer A, leave it running, switch to computer B, answer some pings, and then return to A, you have to kill the politely-waiting `launch.pl` on A, lest it upload its stale log after you close the editor. I may add further checks to avoid this, but killing `launch.pl` is easy enough for me as of now. If you turn off/sleep A, there's nothing to worry about, as it will pull the updated logs when it wakes back up.

Obviously, caveat emptor, make sure you don't upload your logs to publicly-accessible places if you don't want to, I tried to make it as non-destructive as possible but I'm not responsible if it eats your logs and your mother's logs, etc etc.

cincodenada committed:
Copy log to remote location when done

Prepping for some basic syncing

cincodenada committed:
Add basic remote sync functionality

cincodenada committed:
Don't do any remote stuff if remote_id is empty

cincodenada committed:
Hypothetical fix for multiple users

Had to switch to `cd` cause of the &&, but this also now does the `tail`
on the remote server, which saves (a very small amount of) bandwidth

cincodenada committed:
Commit fix for trailing '

Bah, meant to amend this to the last, not sure how it slipped through

cincodenada committed:
Attempt to fill from remote for any gap

`$retrothresh` isn't applicable here, fixes syncing for quick
(<`$retrothresh`) jumps between computers. Should be harmless, if
there's updated pings it'll pull them in, otherwise will fill in or
prompt as normal

Insti commented:
Nice.

cincodenada committed:
Attempt to fix "have to kill launch" issue

Now, closing the window without saving should backfill from remote,
instead of backfilling locally and then uploading

Haven't tested this yet though

cincodenada committed:
Set DISPLAY in launch

This was causing ERR pings when running from systemd

cincodenada committed:
Fix variable name

Missed this when moving into a function

cincodenada committed:
Add systemd service file

cincodenada committed:
Copy all log files, use ping time for most recent

Also fix backtick-and-system nonsense

cincodenada committed:
Better logging

cincodenada committed:
Use tagtime-specific key/user

Also move awk script into its own file and fix silly path error

cincodenada committed:
Add newlines and working directory

cincodenada committed:
Don't poison the generator

cincodenada committed:
Fix remoteln() to also use ping timestamps

Avoids depending on file modification time, which is unreliable

cincodenada committed:
Add a safety before deleting points from a beeminder log

cincodenada committed:
Add more logging

cincodenada committed:
Fix remote_host in template

dreeves commented:
Thanks so much for writing this, @cincodenada! If it's been working nicely for you for the past month I'm sold on merging it!

cincodenada committed:
Disable caching if using remote sync

cincodenada committed:
Fix semicolon

cincodenada committed:
Parameterize scp/ssh commands for overriding

cincodenada committed:
Add initial basic testing harness

cincodenada committed:
Start to deal with time faking, ugh

cincodenada committed:
Move time dispenser into util

cincodenada committed:
Add new variables to template

cincodenada committed:
Merge branch 'remote-sync' of github.com:cincodenada/TagTime into remote-sync

cincodenada committed:
Don't overwrite our own log

cincodenada committed:
Delete all Beeminder points with Y

cincodenada committed:
Make merge a library

cincodenada committed:
Actually merge the tags

cincodenada committed:
Replace fill_remote with call to merge

cincodenada committed:
Remove calculated values from settings.pl

cincodenada committed:
Make a git commit if the remote is a git repo

cincodenada committed:
Fixes

cincodenada committed:
Require util with absolute path

Now that we're using it in the daemon

cincodenada committed:
Log message, commit author

cincodenada committed:
Add filename for NON-MONOTONE

cincodenada committed:
Add debug function that forces a flush

In case we exit/are killed first...this shouldn't be necessary,
but I seem to be dropping log messages, so...

cincodenada committed:
Pop up an editor for merge errors

This shouldn't be a silent failure

cincodenada committed:
Update $nxtping when we pull from remote

This should fix the erroneous err bug

cincodenada committed:
Get more aggressive about failed merges

The previous warning strategy was not effective because
a) the user can't do anything from an editor to resolve this and
b) the error messages weren't going to the file anyway, so it showed an
empty file

So now we just check for the presence of a merge file (if all goes well
it will be cleaned up afterward) and throw up a big ol' warning if it
exists

cincodenada committed:
Give list of files with errors in merge
